HPCC-32735 Optimize replaceString in jstring.cpp

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-32735 Optimize replaceString in jstring.cpp

dcamper commented on system/jlib/jstring.cpp:
```diff
@@ -937,27 +937,58 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
 {
-    if (lenSource)
+    // Only copy if the source and search (oldStr) strings are non-empty
+    if (lenSource && lenOldStr)
```
I think these need to be handled separately. This works for lenSource == 0, but if lenSource > 0 and lenOldStr == 0 then you need to copy the source to the result before returning it.

jackdelv replied:
Fixed. When lenSource > 0 and lenOldStr == 0 the source is copied to result.

dcamper commented on system/jlib/jstring.cpp:
```diff
@@ -937,27 +937,58 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
-                result.append(*source);
-                source++;
+                unmatchedChars++;
+                offset++;
```
Can this else clause repeat the logic of the first while() loop in the function, to compare single characters and increment unmatchedChars without going through memcmp each time?

jackdelv replied:
Yes, I have changed it so it compares single characters until it finds a match to the first character.

dcamper reviewed (changes requested):
A couple of comments.

dcamper commented on system/jlib/jstring.cpp:
```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
+            // If there were any characters left or characters that didn't match, append them
+            left += unmatchedChars;
+            if (left && (offset - left) < lenSource)
+                result.append(left, source + offset - unmatchedChars);
```
This may be clearer as:

result.append(left, source + lenSource - left);

Assuming that what I typed here without checking actually is equivalent; please double-check.

jackdelv replied:
Changed to result.append(left, source + lenSource - left) as lenSource - left = offset - unmatchedChars.

dcamper reviewed (changes requested):
One minor change, for clarity.

dcamper reviewed (approved):
Looks good.

Please check with Gavin to see if it makes sense to copy this algorithm to replaceStringNoCase() in this same PR.

ghalliday commented on system/jlib/jstring.cpp:
```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
+                    left--;
+
+                    // Skip ahead to the next possible match
+                    while (left >= lenOldStr)
```
This is essentially repeating the code at the head of the function.  Better to restructure so that the code only appears once.  E.g.

```
            size_t offset = 0;
            size_t lastCopied = 0;
            size_t maxOffset = lenSource-lenOldStr;
            char firstChar = oldStr[0];
            result.ensureCapacity(lenSource);  // see comment above.
            while (offset <= maxOffset)
            {
                if (source[offset] == firstChar && (memcmp(source+offset, oldStr, lenOldStr) == 0)
                {
                        if (offset != lastCopied)
                            result.append(offset-lastCopied, source + offset);
                        result.append(lenNewStr, newStr);
                        lastCopied = offset + lenOldStr;
                        offset += lenOldStr;
                     }
                 }
                 else
                     offset += 1;
            }
            result.append(lenOldStr - lastCopied, oldStr + lastCopied);
      }
      else....
```
NOTE: This code has not been tested, it was typed in here.

My aim is to make the code as simple (and as clear) as possible, because that is likely to produce the most efficient code.  For example when a character does not match this is only updating a single variable, rather than 3 variables.  It has avoided duplicating the code for scanning.

A further improvement for code that is really time critical is to give the compiler hints about how likely something is.  We have macros likely() and unlikely() which pass through information to the compiler.  So even better:

```
                if (unlikely(source[offset] == firstChar) && 
                    unlikely(memcmp(source+offset, oldStr, lenOldStr) == 0)
```
I am fairly sure this is the default bahavior of the compiler in this case, but it is better to be explicit.

jackdelv replied:
Restructured code and added unlikely macros.

ghalliday commented on system/jlib/jstring.cpp:
```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
+
+            // Reserve space when the new string is longer than the old string
+            unsigned steps = lenSource - lenOldStr - offset + 1;
+            unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
```
You can't guess how long the new string is likely to be, so it is simplest to assume it will match the source.  If there is a match and the replace string is shorter this will be slightly too long, if it is longer it will (possibly) reallocate.  The big advantage in preallocating is to avoid repeated copies as the buffer is built up.  This can become significant for large strings.  A single reallocation by comparison is minor.

jackdelv replied:
This attempts to calculate the max size if for every comparison there is a match. If the lenOldStr<=lenNewStr then it does assume it will match the source. I think this does preallocate enough space so there are no need for reallocations, but if we always assume the length matches the source then sometimes we won't preallocate enough.

ghalliday reviewed (changes requested):
@jackdelv I have added fairly substantial comments.
Normally I wouldn't be concerned about the efficiency.  I would possibly comment on the duplication of the scanning code, but not the rest.
However, because this PR is explicitly to optimize the function I have expanded a lot more.

You could also optimize the case where lenSource==1, avoid the memcmp, (and one other optimization for you to discover).
I would say that is uncommon - but I have just seen a PR that had two calls in a row with single character search strings.

Finally, it needs some unit tests.

See https://github.com/hpcc-systems/HPCC-Platform/pull/19186/commits/f0c3a454fcd3011c90fc77469ce181faef087d61 for a very similar function I just introduced as part of a PR.  (Feel free to spot any bugs in it!)

ghalliday commented:
NOTE: This is called from StringBuffer::replaceString.  That can be further optimized for the two cases where
i) no replacements take place
ii) the source and target replacement strings are the same length
Something for a follow on PR

jackdelv commented on system/jlib/jstring.cpp:
```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            // Scan for match on first character
+            size_t offset = 0;
+            size_t left = lenSource;
```
Fixed

jackdelv commented on system/jlib/jstring.cpp:
```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
     {
-        size_t left = lenSource;
-        while (left >= lenOldStr)
+        if (lenOldStr)
```
Fixed.

jackdelv commented:
@ghalliday Currently, I have a copy of this function in a single .cpp file and I have been running it with a shell script to test. How should I go about implementing Unit Tests?

ghalliday commented on system/jlib/jstring.cpp:
```diff
@@ -937,28 +937,34 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
+        // Reserve space when the new string is longer than the old string
+        unsigned steps = lenSource - lenOldStr + 1;
+        unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
+        result.ensureCapacity(maxResultLength);
```
I'm not convinced this is worthwhile.  This could end up allocating a very large amount of memory, which is unlikely to be used.  The big benefit is to avoid reallocation from 0 up to the size of the source string which would have previously been the case.  I would use

results.ensureCapacity(lenSource);

The calculation is also incorrect it should be
 unsigned [max]steps = lenSource / lenOldStr;

jackdelv replied:
Changed to results.ensureCapacity(lenSource);

ghalliday commented on system/jlib/jstring.cpp:
```diff
@@ -937,28 +937,34 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
+        while (offset < maxOffset)
         {
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            if (unlikely(source[offset] == firstChar) && unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
```
style preference.  I would prefer this split over two lines to make it easier to read.  i.e. check first character, and if that matches compare the rest.

jackdelv replied:
Changed to 
```
 if (unlikely(source[offset] == firstChar) 
    && unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
````

I'm not sure if this is what you meant, but adding an additional if statement would mean I also need another else incrementing output

```
 if (unlikely(source[offset] == firstChar) 
{     
    if (unlikely((lenOldStr == 1) || memcmp(source + offset, oldStr, lenOldStr)==0))
    {
        ...
    }
    else
         offset++;
 }
 else
     offset++;
```

ghalliday reviewed (changes requested):
@jackdelv looks better.
It really needs some unittests (add to JLibStringTest in jlibtests.cpp)

There is at least two bugs that I haven't commented on - see if your unit tests can find them....

jackdelv commented:
@ghalliday I added some unit tests and fixed two bugs. The lenOldStr was equal to 0 and able to get stuck in the loop, and the maxOffset was off by one missing the last match. Were these the two that you mentioned?

ghalliday reviewed (approved):
@jackdelv looks good.  You caught the two problems that I spotted with your test cases.
A good set of test cases.  Please squash.
I have retargeted to master because this is a non-essential optimization, so best to target that where there is no chance of breaking a point release.

dcamper commented:
@ghalliday I think these changes should be applied to replaceStringNoCase() as well. Should that happen in this PR or in a new Jira/PR?

jackdelv commented:
@ghalliday Squashed.

ghalliday commented:
> @ghalliday I think these changes should be applied to replaceStringNoCase() as well. Should that happen in this PR or in a new Jira/PR?

in a new PR - it is much better to keep changes small because they can be merged sooner.
